### PR TITLE
Remove support for python 3.8

### PR DESCRIPTION
**Pull request by ksunden**
## PR Checklist

**Documentation and Tests**
- [ ] Has pytest style unit tests (and `pytest` passes)
- [ ] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).
- [ ] New plotting related features are documented with examples.

**Release Notes**
- [ ] New features are marked with a `.. versionadded::` directive in the docstring and documented in `doc/users/next_whats_new/`
- [ ] API changes are marked with a `.. versionchanged::` directive in the docstring and documented in `doc/api/next_api_changes/`
- [ ] Release notes conform with instructions in  `next_whats_new/README.rst` or `next_api_changes/README.rst`

### CODE REVIEW

File: .github/workflows/tests.yml [hunk 1]
```diff
@@ -37,17 +37,13 @@ jobs:
...
+            pyqt5-ver: '==5.12.2 sip==5.0.0'  # oldest versions with a Py3.9 wheel.
             pyqt6-ver: '==6.1.0 PyQt6-Qt6==6.1.0'
-            pyside2-ver: '==5.14.0'  # oldest version with working Py3.8 wheel.
+            pyside2-ver: '==5.15.1'  # oldest version with working Py3.9 wheel.
             pyside6-ver: '==6.0.0'
             delete-font-cache: true
-          - os: ubuntu-20.04
-            python-version: 3.8
-            extra-requirements: '-r requirements/testing/extra.txt'
-            CFLAGS: "-fno-lto"  # Ensure that disabling LTO works.
```
Comment (tacaswell) [inline .github/workflows/tests.yml:50]: We probably want to keep the no LTO test case?

**Comment by tacaswell** [comment]
The minimum test is failing because we are building pillow from source with out webp, we probably should add a skip check on that test (I have a vague memory of someone else reporting the same failure, but not sure where).

**Comment by ksunden** [comment]
Alternatively, we could bump the minimum version of Pillow if we wanted to ensure wheels are available.

Looks like the first version to supply a 3.9 wheel is pillow 8.0.0 (released Oct 2020)

By NEP 29/SPEC 0 standards (24 months) would be pillow 8.1 or thereabouts (and later for expected mpl 3.8 release date)

File: .github/workflows/tests.yml (hunk 1 above)
Comment (ksunden) [reply .github/workflows/tests.yml:50]: added back (to 3.9)

**Comment by oscargus** [comment]
webp was here: https://github.com/matplotlib/matplotlib/pull/24711 (Py 3.12)

File: setup.py [hunk 2]
```diff
@@ -8,7 +8,7 @@
 # and/or pip.
 import sys
 
-py_min_version = (3, 8)  # minimal supported python version
+py_min_version = (3, 9)  # minimal supported python version
 since_mpl_version = (3, 6)  # py_min_version is required since this mpl version
```
Comment (oscargus) [inline setup.py:12]:
```suggestion
since_mpl_version = (3, 8)  # py_min_version is required since this mpl version
```
?

**Comment by ksunden** [comment]
By Pillow's [docs](https://pillow.readthedocs.io/en/stable/installation.html#id2), they do not support pillow <8.0 on py3.9 so I think I'll start there. I don't disagree that a graceful fallback/skip/xfail is a better solution long term, just lean towards following upstream when it comes to python support.

If it still fails, I'll add the skips here, but if it passes, I think that can be done in an alternate PR (either on its own or in #24711)

**Comment by ksunden** [comment]
Other potential bump to minimum supported versions (picked out because the minimum version currently specified is not installable as a wheel on 3.9):

- kiwisolver: bump from 1.0.1 (rel Oct 2017) to 1.3.1 (rel Nov 2020, 1.3.0 has py3.9 wheel, but was marked as yanked on pypi)

**Review by tacaswell — APPROVED** [review]

**Comment by ksunden** [comment]
Appveyor may need to be 3.10, waiting for the queue to see if the problems on 3.11 persist

**Comment by oscargus** [comment]
Failure comes from GTK4-warnings: https://gitlab.gnome.org/GNOME/pygobject/-/issues/494

**Comment by tacaswell** [comment]
It looks likes we already filter warnings on that test and just need to add `ImportWarning` as well (which I am going to assume is the next step up from the DeprecationWarning that gtk4 was triggering?)

File: doc/devel/dependencies.rst [hunk 3]
```diff
@@ -15,15 +15,15 @@ When installing through a package manager like ``pip`` or ``conda``, the
...
 
-* `Python <https://www.python.org/downloads/>`_ (>= 3.8)
+* `Python <https://www.python.org/downloads/>`_ (>= 3.9)
 * `contourpy <https://pypi.org/project/contourpy/>`_ (>= 1.0.1)
 * `cycler <https://matplotlib.org/cycler/>`_ (>= 0.10.0)
 * `dateutil <https://pypi.org/project/python-dateutil/>`_ (>= 2.7)
 * `fontTools <https://fonttools.readthedocs.io/en/latest/>`_ (>= 4.22.0)
-* `kiwisolver <https://github.com/nucleic/kiwi>`_ (>= 1.0.1)
+* `kiwisolver <https://github.com/nucleic/kiwi>`_ (>= 1.3.1)
 * `NumPy <https://numpy.org>`_ (>= 1.20)
```
Comment (oscargus) [inline doc/devel/dependencies.rst:24]: Bump to 1.21 while at it? (End of month)

**Review by oscargus — APPROVED** [review]
Feel free to self merge if you do not want to bump numpy now.

**Comment by oscargus** [comment]
Did the Azure tests stop working with this PR or was it just around this PR (or am I missing something)?

**Comment by EwoutH** [comment]
[NEP 29](https://numpy.org/neps/nep-0029-deprecation_policy.html) was superseded by the scientific python ecosystem coordination guideline [SPEC 0 — Minimum Supported Versions](https://scientific-python.org/specs/spec-0000/). Where NEP had a 42 month support windows (3.5 years) for each Python version, SPEC 0 recommends a 3 year support window.

If we were to follow SPEC 0, than Python 3.9 support can already be dropped, since it was released October 5th, 2020.

**Comment by QuLogic** [comment]
3.8 is already released; we aren't going to change the supported Python version in it.

**Comment by EwoutH** [comment]
Not dropping Python 3.8 support for the existing Matplotlib 3.8.x releases of course, but it can be dropped on the current master branch (towards the 3.9.x releases).

(which coincidentally is very fitting, requiring Python 3.9 for Matplotlib 3.9)

**Pull request merged**